core/server/master/src/main/java/alluxio/master/backup/BackupLeaderRole.java

```diff
@@ -0,0 +1,293 @@
+import alluxio.resource.LockResource;
+import alluxio.security.authentication.ClientIpAddressInjector;
+import alluxio.util.ConfigurationUtils;
+import alluxio.util.network.NetworkAddressUtils;
+import alluxio.wire.BackupStatus;
+
+import io.atomix.catalyst.transport.Address;
+import io.atomix.catalyst.transport.Connection;
+import io.grpc.ServerInterceptors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.function.Consumer;
+
+/**
+ * Implementation of {@link BackupRole} for primary mode.
+ */
+public class BackupLeaderRole extends AbstractBackupRole {
+  private static final Logger LOG = LoggerFactory.getLogger(BackupLeaderRole.class);
+
+  // Timeout after which backup will be abandoned if no heart-beat received.
+  private final long mBackupAbandonTimeout;
+
+  /** Metadata state pause lock. */
+  private Lock mStatePauseLock;
+
+  /** Scheduled future to time-put backups on leader. */
+  private ScheduledFuture<?> mTimeoutBackupFuture;
+  /** Time at which the last heart-beat was received by leader. */
+  private Instant mLastHeartBeat;
+
+  /** Connection/Future pair to control ongoing backup. */
+  private Pair<Connection, Future<?>> mBackupControl;
+
+  /** Follower master connections with the leader. */
+  private Set<Connection> mFollowerConnections = new ConcurrentHashSet<>();
+
+  /**
+   * Creates a new backup leader.
+   *
+   * @param masterContext the master context
+   */
+  public BackupLeaderRole(CoreMasterContext masterContext) {
+    super(masterContext);
+    // Store state lock for pausing state change when necessary.
+    mStatePauseLock = masterContext.pauseStateLock();
+    // Read properties.
+    mBackupAbandonTimeout =
+        ServerConfiguration.getMs(PropertyKey.MASTER_BACKUP_ABANDON_TIMEOUT_MS);
+  }
+
+  @Override
+  public void start() throws IOException {
+    LOG.info("Starting backup leader role.");
+  }
+
+  @Override
+  public void stop() throws IOException {
+    LOG.info("Stopping backup leader role.");
+    // Reset existing stand-by connections.
+    mFollowerConnections.clear();
+    // Cancel ongoing backup task.
+    if (mBackupControl != null && mBackupControl.getSecond() != null) {
+      mBackupControl.getSecond().cancel(true);
+    }
+    // Stopping the base after because closing connection uses the base executor.
+    super.stop();
+  }
+
+  @Override
+  public Map<ServiceType, GrpcService> getRoleServices() {
+    Map<ServiceType, GrpcService> services = new HashMap<>();
+    services
+        .put(ServiceType.META_MASTER_BACKUP_MESSAGING_SERVICE,
+            new GrpcService(
+                ServerInterceptors.intercept(
```

Comment:
`ClientIpAddressInjector` is being injected to every call into`CopycatMessageServiceClientHandler` here. 

It's used mostly for traces. Otherwise It's very hard to say who's connecting to server in a gRPC bi-di stream.

Reply:
Got it, thanks